server/text_generation_server/layers/attention/kv_cache.py

```diff
@@ -52,13 +52,23 @@ def __init__(
         device: torch.device,
     ):
         """Construct the key-value cache for a layer."""
-
-        if dtype in {torch.float8_e5m2, torch.float8_e4m3fn} and (
-            ATTENTION != "flashinfer" or SYSTEM != "cuda"
-        ):
-            raise ValueError(
-                "FP8 KV cache is currently only supported for flashinfer on CUDA"
-            )
+        if dtype in {torch.float8_e5m2, torch.float8_e4m3fn}:
+            if not (
+                (ATTENTION == "flashinfer" and SYSTEM == "cuda")
+                or (ATTENTION == "paged" and SYSTEM == "rocm")
+            ):
+                raise ValueError(
+                    "FP8 KV cache is currently only supported for flashinfer on CUDA and paged attention on ROCm. "
+                )
+            if SYSTEM == "rocm" and dtype == torch.float8_e5m2:
+                raise ValueError(
+                    "float8_e5m2 FP8 KV cache is not supported on AMD ROCm"
+                )
+
+        self.kv_cache_dtype = "auto"
+        if SYSTEM == "rocm" and dtype == torch.float8_e4m3fn:
+            self.kv_cache_dtype = "fp8"
+            dtype = torch.uint8
```

Comment:
I think it would be nicer to allocate the cache in the requested type (`torch.float8_e4m3fn`) and then cast to `uint8` where/when needed. Then we don't need another bookkeeping field (`kv_cache_dtype`) and the cache carries the type information.

Reply:
Updated